solarforecastarbiter/reference_forecasts/main.py

```diff
@@ -61,18 +70,179 @@ def run(site, model, init_time, start, end):
     else:
         ac_power = None
 
     # resample data after power calculation
-    resampled = list(map(resampler, (*solpos_forecast, ac_power)))
+    resampled = list(map(resampler, (*forecast, ac_power)))
     return resampled
+
+
+def run_persistence(observation, forecast, issue_time, index=False):
+    """
+    Run a persistence *forecast* for an *observation*.
+
+    For intraday forecasts, the *index* argument controls if the
+    forecast is constructed using persistence of the measured values
+    (*index = False*) or persistence using clear sky index or AC power
+    index.
+
+    For day ahead forecasts, only persistence of measured values
+    (*index = False*) is supported.
+
+    The persistence *window* is the time over which the persistence
+    quantity (irradiance, power, clear sky index, or power index) is
+    averaged over. The persistence window is automatically determined
+    based on the type of persistence forecast desired:
+
+      * Intraday persistence forecasts: Maximum of observation interval
+        length, forecast interval length, and forecast lead time to
+        start. No longer than 1 hour.
+      * Day ahead forecasts: Equal to forecast interval length.
+
+    Users that would like more flexibility may use the lower-level
+    functions in
+    :py:mod:`solarforecastarbiter.reference_forecasts.persistence`.
+
+    Parameters
+    ----------
+    observation : datamodel.Observation
+        The metadata of the observation to be used to create the
+        forecast.
+    forecast : datamodel.Forecast
+        The metadata of the desired forecast.
+    issue_time : pd.Timestamp
+        Issue time of the forecast run.
+    index : bool, default False
+        If False, use persistence of observed value. If True, use
+        persistence of clear sky or AC power index.
+
+    Returns
+    -------
+    forecast : pd.Series
+        Forecast conforms to the metadata specified by the *forecast*
+        argument.
+
+    Raises
+    ------
+    ValueError
+        If forecast and issue_time are incompatible.
+    ValueError
+        If forecast.run_length >= 1 day and index=True.
+    ValueError
+        If instantaneous forecast and instantaneous observation interval
+        lengths do not match.
+    ValueError
+        If average observations are used to make instantaneous forecast.
+    """
+    forecast_start, forecast_end = get_forecast_start_end(forecast, issue_time)
+
+    # intra day persistence and "day ahead" persistence require
+    # fairly different parameters.
+    # is this a sufficiently robust way to distinguish?
+    intraday = forecast.run_length < pd.Timedelta('1d')
+    if intraday:
+        # time window over which observation data will be used to create
+        # persistence forecast.
+        # assumes observation.interval_length <= 1h
+        window = min(
+            pd.Timedelta('1h'),
+            max(observation.interval_length, forecast.interval_length,
```

Comment:
This is a bit confusing.  What cases came up where `window = pd.Timedelta('1hr')` wouldn't work?

Reply:
`window = pd.Timedelta('1hr')` does work and is in fact what `window` is set to for the relatively common scenario that `forecast.lead_time_to_start` or `forecast.run_length` is > 1hr.

Reply:
I am suggesting that the `max` part of the window determination may be unnecessarily complicating matters.

Reply:
I'm trying to think alternatives:

1. Always use `window = 1h`. Arguably a bad approach for persistence forecasts with intervals and duration << 1h.
2. Expand the definition of a `datamodel.Forecast` so that it's predefined there. Feels like the wrong place because the `datamodel.Forecast` currently says nothing about implementation.
3. Add a `kwarg`. What's the default value? Does the script that calls `run_persistence` inspect the `observation` and `forecast` to change the default value?

I can certainly move this logic into its own function(s) that return `data_start, data_end` and that might help clarify the overall `run_persistence` function.

Reply:
Agree that the helper function would improve this code.

1. OK.  If we envision a 15min forecast with 5min intervals, agree that window = 1hr probably isn't helpful.  But it's a reference forecast, so it's more important that the forecast method is transparent and repeatable, than 'best' in some sense, in my view.

What about `min(forecast.run_length, pd.Timedelta('1hr'))`?  I'd guess that run_length would tend to be the longest of the 4 quantities in the current line.

Reply:
Agree with prioritizing transparent and repeatable. What happens if `forecast.run_length` < `observation.interval_length`? Error? Or maybe `window = max(observation.interval_length, forecast.run_length)` and forget about the `1h` limit?

Reply:
> What happens if forecast.run_length < observation.interval_length? Error? 

I would, basically that asks the forecast to downscale from the observations.  I have a bias to think of the persistence index forecast as a forecast of a mean, and so hesitate to apply the method for short times, e.g., less than 15 min, where a passing shadow will dominate the observations.